<h1>3. Skill Prediction en LinkedIN

**3.0 Importacines necesarias**

Se importan librerías y métodos necesarios para el correcto funcionamiento de las activades a desarrollar dentro de ésta sección.

In [27]:
#!usr/bin/env python
# -*- coding: utf-8 -*-
from scipy.sparse import csr_matrix
import re
import numpy as np
from sklearn.lda import LDA
from sklearn.qda import QDA

**3.1 Construcción de dataset**

Primeramente, se crea la función *generar_Z*, la cual, a partir de los archivos *user_skill* y *skill_id*, construye una matriz dispersa con los datos provistos en dicho archivo. Tal matriz será denotada de aquí en adelante como Z.

In [3]:
#Se define funcion para construir matriz Z
def generar_Z(arch1, arch2):
    #Variables que contienen la cantidad de perfiles y de competencias, respectivamente
    n_perfiles = 0
    n_competencias = 0
    
    #Se determina la cantidad de perfiles
    with open(arch1) as perfiles:
        n_perfiles = sum(1 for _ in perfiles)
    
    #Se determina la cantidad de competencias
    with open(arch2) as competencias:
        n_competencias = sum(1 for _ in competencias)
        
    #Se crea fisicamente la matriz Z
    Z = csr_matrix((n_perfiles, n_competencias), dtype=np.int8).toarray()
    
    #Se procede a 'trasladar' los valores de los archivos a Z
    with open(arch1) as perfiles:
        for line in perfiles:
            datos = re.split(':|,|\n', line)
            n_perfil = int(datos[0])
            for n_competencia in datos[1:-1]:
                Z[n_perfil][int(n_competencia)] = 1
    
    return Z

Luego, se procede a construir la matriz Z con que se trabajará de ahora en adelante.

In [4]:
Z = generar_Z('user_skill', 'skill_id')

A partir del resultado anterior, se puede obtener la cantidad de perfiles y de competencias que integran los dataset.

In [5]:
print 'Se cuenta con un dataset compuesto por', Z.shape[0], 'perfiles y', Z.shape[1], 'competencias.'

Se cuenta con un dataset compuesto por 7890 perfiles y 14544 competencias.


**3.2 Construcción de set de entrenamiento y set de pruebas**

A partir de la matriz Z generada en la sección anterior, se procede a construir los conjuntos de entrenamiento y de prueba. Para ello, se seleccionan registros desde Z en forma azarosa de tal forma que el conjunto de entrenamiento estará compuesto por el 60% de los registros del dataset original, mientras que el conjunto de prueba estará integrando por el 40% de los registros de dicho dataset. Considerando que Z acumula 7890 perfiles, el conjunto de entrenamiento tendrá 4734 registros y el de prueba, 3156 registros.

Para realizar la tarea descrita, se sigue el procedimiento presentado a continuación:

In [6]:
#Se permutan aleatoriamente las filas de la matriz Z
np.random.shuffle(Z)

#Se construyen conjunto de entrenamiento y de prueba, respectivamente
Z_tr = Z[0:int(0.6*Z.shape[0])]
Z_ts = Z[int(0.6*Z.shape[0]):]

**3.3 Cantidad de perfiles por competencia**

En ésta sección, se desea construir un gráfico en el que sea posible apreciar la cantidad de perfiles que declaran una determinada competencia. Para esto, se lleva a cabo el siguiente procedimiento:

In [8]:
#Se obtiene el nombre, identificador y cantidad de perfiles asociado a cada competencia, y se 
#almacenan en una lista de tuplas

lista_competencias = []
with open('skill_id') as competencias:
    for line in competencias:
        dato = re.split(' :|:|\n', line)
        dato = dato[0:-1]
        
        if (len(dato) > 2):
            nombre_competencia = ''
            for token in dato[0:-1]:
                nombre_competencia += token
            id_competencia = int(dato[-1])
            cantidad_perfiles = sum(Z[:, id_competencia])
            lista_competencias.append((nombre_competencia, id_competencia, cantidad_perfiles))
            
        else:
            if (dato[0] != ''):
                nombre_competencia = dato[0]
                id_competencia = int(dato[1])
                cantidad_perfiles = sum(Z[:, id_competencia])
                lista_competencias.append((nombre_competencia, id_competencia, cantidad_perfiles))
            
#Finalizado el proceso anterior, se ordena la lista competencias por su frecuencia
lista_competencias.sort(key=lambda tup: tup[2], reverse=True)

#A continuacion, se muestra un ranking con las 10 competencias mas frecuentes
print('A continuación, se muestra un ranking con las 10 competencias más frecuentes (con su id entre paréntesis), acompañadas por la cantidad de perfiles en que es declarada. Es importante considerar que sólo se han tomado en \ncuenta aquellas competencias cuyo nombre es no nulo.\n')
contador = 0
for comp in lista_competencias:
    print contador+1,'.', comp[0], '(', comp[1],')',':', comp[2]
    contador += 1
    if (contador > 9):
        break

A continuación, se muestra un ranking con las 10 competencias más frecuentes, acompañadas por la cantidad de perfiles en que es declarada. Se muestran en el formato ranking. Competencia (id_competencia) : frecuencia. Es importante considerar que sólo se han tomado en 
cuenta aquellas competencias cuyo nombre es no nulo.

1 . Microsoft Office ( 29 ) : 3081
2 . Liderazgo de equipos ( 32 ) : 2761
3 . Gestión de proyectos ( 43 ) : 2204
4 . Microsoft Excel ( 71 ) : 2191
5 . Team Leadership ( 1 ) : 2109
6 . Estrategia empresarial ( 47 ) : 1868
7 . SQL ( 55 ) : 1782
8 . Planificación estratégica ( 45 ) : 1717
9 . Inteligencia empresarial ( 49 ) : 1673
10 . Inglés ( 33 ) : 1647


**3.4 Predicción de la competencia *Microsoft Office***

**3.4.1 Ajuste de conjuntos de entrenamiento y prueba**

Se desea predecir la existencia de la competencia *Microsoft Office* en un perfil determinado, en base a las competencias restantes que el usuario declaró. Se ha seleccionado dicha competencia debido a su alta frecuencia de aparición. Para el posterior entrenamiento/ajuste de clasificadores, se ajustan los sets de entrenamiento y de prueba construídos previamente para así poder abordar este problema. Se procede de la siguiente manera:

In [21]:
#Para el set de entrenamiento, se construyen las matrices X_tr e y_tr a partir de Z_tr
#Recordar que Microsoft Office corresponde a la competencia numero 29
y_tr = Z_tr[:, 29]
X_tr = np.delete(Z_tr, 29, 1)

#Se procede de forma analoga con el conjunto de entrenamiento, construyendo X_ts e y_ts
#a partir de Z_ts
y_ts = Z_ts[:, 29]
X_ts = np.delete(Z_ts, 29, 1)

**3.4.2 Entrenamiento/ajuste de clasificador LDA**

In [ ]:
lda_model = LDA()
lda_model.fit(X_tr, y_tr)
score_tr = lda_model.score(X_tr, y_tr)
score_ts = lda.model.score(X_ts, y_ts)